src/axom/core/tests/core_array.hpp

```diff
@@ -2257,3 +2257,79 @@ TEST(core_array, resize_stackarray)
   test_resize_with_stackarray<bool>(false);
   test_resize_with_stackarray<int>(-1);
 }
+
+//------------------------------------------------------------------------------
+constexpr static int NONTRIVIAL_RELOC_MAGIC = 123;
+struct NonTriviallyRelocatable
+{
+  NonTriviallyRelocatable()
+    : m_member(NONTRIVIAL_RELOC_MAGIC)
+    , m_localMemberPointer(&m_member)
+  { }
+
+  NonTriviallyRelocatable(const NonTriviallyRelocatable& other)
+    : m_member(other.m_member)
+    , m_localMemberPointer(&m_member)
+  { }
+
+  NonTriviallyRelocatable& operator=(const NonTriviallyRelocatable& other)
+  {
+    if(this != &other)
+    {
+      m_member = other.m_member;
+    }
+    return *this;
+  }
+
+  ~NonTriviallyRelocatable() = default;
+
+  int m_member;
+  int* m_localMemberPointer;
+};
+
+TEST(core_array, reserve_nontrivial_reloc)
```

Comment:
Do these tests cover all the cases of different allocators (`HOST`, `DEVICE`, `DYNAMIC`)?

It looks like it only covers the CPU malloc allocator, but I might be missing something.

Reply:
Perhaps it doesn't matter since the copies are only allowed on the host (?), but there might be a case where we're using a custom allocator on the host? E.g. a pooled allocator, or a shared-memory allocator?

Reply:
I'll add a test for unified memory.

For device memory, we don't currently support non-trivially relocatable objects, since we'd need to be able to invoke the copy/move constructor in a device kernel. Instead of forcing users to annotate all their constructors as host/device, we pessimistically assume that non-trivial constructors are host-only.

In the future, we might consider allowing users to "declare" via a type trait specialization that their user-defined type is constructible/destructible on the host, which would allow us to support this use case for device memory.